src/Server/ILanguageServerConfiguration.cs

```diff
@@ -0,0 +1,38 @@
+﻿using System;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Configuration;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using OmniSharp.Extensions.LanguageServer.Server.Configuration;
+
+namespace OmniSharp.Extensions.LanguageServer.Server
+{
+    public interface ILanguageServerConfiguration : IConfiguration
+    {
+        /// <summary>
+        /// Gets the current configuration values from the client
+        /// This configuration object is stateless such that it won't change with any other configuration changes
+        /// </summary>
+        /// <param name="items"></param>
+        /// <returns></returns>
+        Task<IConfiguration> GetConfiguration(params ConfigurationItem[] items);
+
+        /// <summary>
+        /// Gets the current configuration for a given document uri
+        /// This re-uses all the sections from the <see cref="ConfigurationItem"/>s that
+        /// the root configuration uses.
+        ///
+        /// This will watch for changes of the scoped documents and update the configuration.
+        /// </summary>
+        /// <param name="scopeUri"></param>
+        /// <returns></returns>
+        Task<IDisposableConfiguration> GetScopedConfiguration(Uri scopeUri);
```

Comment:
Hmm, is there value in exposing that this is an IDisposableConfiguration here? Is the LanguageServer author expected to call dispose on it?

Reply:
Yeah, the high level idea here is that a language server wants to get configuration information for a specific document (at least from the vscode side anyway).  This would allow them to see any overrides that might have happened for example vscode lets you do `"[csharp]":{}` to override settings for a specific language.

Since there is nothing in the protocol today that lets us ask for settings for a generic language or file type, we instead have to go this route.   The reason it's disposable is because we don't want to have to double information for files that are no longer in scope, as each refresh will update all the active scopes.

Reply:
I understand that aspect of it but by exposing the IDisposeableness of the scoped configuration I as the Razor langauge server author could call dispose which wouldn't be expected I'd imagine. I'd think you (as O# lib) would want to reserve the right to dispose

Reply:
The problem is on the LSP side I don't know when you're "done" with that configuration, hence the disposable aspect of it.

It could be in context of a file or it could be in context of a file that is currently being acted upon.  In essence you wanted the configuration for that file, and you take ownership of when you're done. On the library side we're only going to update the configuration for you automagically.

I could be something that is unnecessary and maybe `GetConfiguration` is enough.

Reply:
Ahh I see